Don't resolve symlinks in -Wconf:src

mbland opened the pull request:
Updates the -Wconf:src filter to resolve each source path to its absolute, normalized form (instead of its canonicalized form) before matching it.

Previously, filters intended to match symlinked paths would not apply when their canonicalized paths didn't contain the same path components. For example, Bazel 9 changed the location of symlinked source repositories for external dependencies such that their canonicalized paths no longer contained an `external/` component. This rendered `-Wconf:src=external/.*:s` filters to silence warnings in these external source repositories ineffective.

This implementation closely matches the Scala3 implementation (as of 3.8.0-RC3), though it does not use `toURI`. Testing revealed that `toURI` prepends the current working directory to Windows-like paths unconditionally, and converts backslashes in such paths to `%5C` escape sequences.

- https://github.com/scala/scala3/blob/3.8.0-RC3/compiler/src/dotty/tools/dotc/reporting/WConf.scala#L33

Fixes scala/bug#13145. Review by @lrytz.

mbland committed:
Don't resolve symlinks in -Wconf:src

Updates the -Wconf:src filter to resolve each source path to its
absolute, normalized form (instead of its canonicalized form) before
matching it.

Previously, filters intended to match symlinked paths would not apply
when their canonicalized paths didn't contain the same path components.
For example, Bazel 9 changed the location of symlinked source
repositories for external dependencies such that their canonicalized
paths no longer contained an `external/` component. This rendered
`-Wconf:src=external/.*:s` filters to silence warnings in these external
source repositories ineffective.

This implementation closely matches the Scala3 implementation (as of
3.8.0-RC3), though it does not use `toURI`. Testing revealed that
`toURI` prepends the current working directory to Windows-like paths
unconditionally, and converts backslashes in such paths to `%5C` escape
sequences.

- https://github.com/scala/scala3/blob/3.8.0-RC3/compiler/src/dotty/tools/dotc/reporting/WConf.scala#L33

lrytz commented on src/compiler/scala/tools/nsc/Reporting.scala:
```diff
@@ -774,7 +774,8 @@ object Reporting {
       def check(pos: Position) = cache.getOrElseUpdate(pos.source, {
-        val sourcePath = pos.source.file.canonicalPath.replace("\\", "/")
+        val sourceFile = pos.source.file.absolute.file
+        val sourcePath = sourceFile.toPath.normalize.toString.replace("\\", "/")
```
`source.file` is an `AbstractFile`, which has a bunch of subtypes. We need to make sure what we do here works / is safe for all of them; in particular, `VirtualFile.file` seems to be `null`.

Sad that this didn't turn up in our test suite...

```suggestion
        val sourcePath = Option(sourceFile).map(_.toPath.normalize.toString).getOrElse(pos.source.path).replace("\\", "/")
```

mbland replied:
Done in ce5c0c68cfd7dca2e7cd838cbd5e95fb3ff8d515. Added `WConfTest.sourcePatternMatchesVirtualFiles` to validate the before and after behavior.

mbland committed:
Ensure -Wconf:src filter handles VirtualFiles

If the `AbstractFile` returned by `pos.source.file` is a `VirtualFile`,
invoking `.absolute.file` on it will return `null`. In that case, we
apply the filter to `pos.source.path` instead.

lrytz reviewed (approved):
Thank you!

mbland commented:
> Is that something we should fix on the Scala 3 side?

I can try it today and report back.

mbland commented:
@lrytz I just filed scala/scala3#24771 and scala/scala3#24772 to avoid `Path.toURI` and to fix Windows paths. I also adapted some of the test cases from this pull request.

lrytz commented:
Thank you!
